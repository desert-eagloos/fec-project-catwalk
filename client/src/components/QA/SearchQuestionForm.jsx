import React, { useState, useEffect, } from 'react';
import { Form, FormControl } from 'react-bootstrap'


const SearchQuestionForm = (props) => {

<<<<<<< HEAD
  // console.log(props);
=======
>>>>>>> develop
  const [searchBarVal, setSearchBarVal] = useState('');

  const updateSearchBarVal = (e) => {
    setSearchBarVal(e.target.value)
  }

  useEffect (() => {
    if (searchBarVal.length > 2) {
      let filteredQuestions = props.originalData.results.filter((question) => {
        return question.question_body.toLowerCase().indexOf(searchBarVal.toLowerCase()) !== -1;
      })
      props.setData({
        "product_id": props.data.product_id,
        "results": filteredQuestions
      })
    } else if (props.originalData && searchBarVal.length < 3) {
      let filteredQuestions = props.originalData.results.slice();
      props.setData({
        "product_id": props.originalData.product_id,
        "results": filteredQuestions
      })
    }

  }, [searchBarVal])

  return (
    <Form>
      <FormControl type='tesxt' placeholder='Have a question? Search for answersâ€¦' onChange={(e) => updateSearchBarVal(e)}/>
    </Form>
  )
}

export default SearchQuestionForm;