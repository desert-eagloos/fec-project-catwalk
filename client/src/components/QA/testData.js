var testData = [
  {
    question: 'Can I throw it in the ocean?',
    answers: [
      'I did and it works just fine.',
      'Not very aerodynamic.',
      'I threw it farther than anything I have ever thrown'
    ]
  },
  {
    question: 'Does this product work in space?',
    answers: [
      'I took it to Mars last week. Works great.',
      'Aliens wont use it but great for humans.'
    ]
  },
  {
    question: 'Will I need show this to my friend Tom?',
    answers: [
      'Under no circumstances show this to Tom.'
    ]
  },
  {
    question: 'Should I be prepared for the zombie apocalypse?',
    answers: [
      'Always',
      'Anything can happen in 2021.',
      'Rule number 1... cardio.'
    ]
  },
  {
    question: 'What gives you the right?',
    answers: [

    ]
  },
]

export default testData;