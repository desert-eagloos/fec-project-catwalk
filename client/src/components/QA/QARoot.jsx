import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';
import { Container, Col, Row } from 'react-bootstrap';
import axios from 'axios';
import QuestionList from './QuestionList';
import SearchQuestionForm from './SearchQuestionForm';

const _ = require('underscore');

const QARoot = ({ product }) => {
  const [id, setID] = useState(product.id);
  const [data, setData] = useState();
  const [originalData, setOriginalData] = useState();
  const [isLoading, setIsLoading] = useState(false);

  const restructureResponseData = (resData) => ({
    product_id: resData.product_id,
    results: _.map(resData.results, (answers) => ({
      question_id: answers.question_id,
      questions_body: answers.question_body,
      questions_date: answers.question_date,
      asker_name: answers.asker_name,
      question_helpfulness: answers.question_helpfulness,
      reported: answers.reported,
      answers: _.map(answers.answers, (answer) => ({
        id: answer.id,
        body: answer.body,
        date: answer.date,
        answerer_name: answer.answerer_name,
        helpfulness: answer.helpfulness,
        photos: answer.photos,
      })),
    })),
  });

  useEffect(() => {
    const asyncFunc = async () => {
      const qaGetReq = async (idp) => axios.get(`/qa/questions/${idp}`);
      const qaGetResponse = await qaGetReq(product.id);
      return qaGetResponse.data;
    };

    asyncFunc()
      .then((res) => {
        const restructuredData = restructureResponseData(res);
        // console.log(restructuredData);
        setOriginalData(restructuredData);
        setData(restructuredData);
        setID(Number(restructuredData.product_id));
        setIsLoading(true);
      })
      .catch();
  }, [product]);

  const renderQuestionList = () => {
    if (isLoading) {
      return (
        <QuestionList data={data} />
      );
    }
    return (<></>);
  };

  if (id === null) {
    return (<></>);
  }

  return (
    <div>
      <Container fluid>
        <Row>
          <Col>
            <h3 id="title">Questions and Answers</h3>
            {originalData
            && <SearchQuestionForm originalData={originalData} data={data} setData={setData} />}
            {renderQuestionList()}
          </Col>
        </Row>
      </Container>
    </div>
  );
};

QARoot.propTypes = {
  product: PropTypes.shape({
    id: PropTypes.number,
    campus: PropTypes.string,
    slogan: PropTypes.string,
    description: PropTypes.string,
    category: PropTypes.string,
    name: PropTypes.string,
    default_price: PropTypes.string,
    created_at: PropTypes.string,
    updated_at: PropTypes.string,
  }),
};

QARoot.defaultProps = {
  product: {
    id: 18080,
    campus: 'hr-bld',
    slogan: 'Odit dolorem nemo id tempora qui.',
    description: 'A sapiente hic. Facilis et sit voluptatem. Ex sunt reiciendis qui ut perferendis qui soluta quod.',
    category: 'Sweatpants',
    name: 'Ernesto Sweatpants',
    default_price: '56.00',
    created_at: '2021-02-23T05:08:00.520Z',
    updated_at: '2021-02-23T05:08:00.520Z',
  },
};

export default QARoot;
