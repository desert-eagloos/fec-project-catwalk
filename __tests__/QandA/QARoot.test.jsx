import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import QARoot from '../../client/src/components/QA/QARoot';
import SearchQuestionForm from '../../client/src/components/QA/SearchQuestionForm';
import testData from '../../client/src/components/QA/testData';

describe('Questions and Answer widget should render', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<QARoot />);
  });
  test('Should render a title', () => {
    console.log(wrapper);
    expect(wrapper.find('#title').text()).toBe('Questions and Answers');
  });
});

describe('Creates Question Search Bar', () => {
  test('Question Search Bar', () => {
    const component = renderer.create(<SearchQuestionForm />);
    const json = component.toJSON();
    expect(json).toMatchSnapshot();
  });
});