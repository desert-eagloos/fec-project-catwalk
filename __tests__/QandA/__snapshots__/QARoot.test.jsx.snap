// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Creates Question Search Bar Question Search Bar 1`] = `
<form
  className=""
>
  <input
    className="form-control"
    onChange={[Function]}
    placeholder="Have a question? Search for answersâ€¦"
    type="tesxt"
  />
</form>
`;
